src/torchpme/tuning/ewald.py

```diff
@@ -0,0 +1,196 @@
+import torch
+
+from ..calculators import EwaldCalculator
+from ..utils import _validate_parameters
+from .tuner import GridSearchTuner, TuningErrorBounds
+
+
+def tune_ewald(
+    charges: torch.Tensor,
+    cell: torch.Tensor,
+    positions: torch.Tensor,
+    cutoff: float,
+    exponent: int = 1,
+    neighbor_indices: Optional[torch.Tensor] = None,
+    neighbor_distances: Optional[torch.Tensor] = None,
+    ns_lo: int = 1,
+    ns_hi: int = 14,
+    accuracy: float = 1e-3,
+) -> tuple[float, dict[str, float]]:
+    r"""
+    Find the optimal parameters for :class:`torchpme.EwaldCalculator`.
+
+    The error formulas are given `online
+    <https://www2.icp.uni-stuttgart.de/~icp/mediawiki/images/4/4d/Script_Longrange_Interactions.pdf>`_
+    (now not available, need to be updated later). Note the difference notation between
+    the parameters in the reference and ours:
+
+    .. math::
+
+        \alpha &= \left( \sqrt{2}\,\mathrm{smearing} \right)^{-1}
+
+        K &= \frac{2 \pi}{\mathrm{lr\_wavelength}}
+
+        r_c &= \mathrm{cutoff}
+
+    :param charges: torch.Tensor, atomic (pseudo-)charges
+    :param cell: torch.Tensor, periodic supercell for the system
+    :param positions: torch.Tensor, Cartesian coordinates of the particles within the
+        supercell.
+    :param cutoff: float, cutoff distance for the neighborlist
+    :param exponent: :math:`p` in :math:`1/r^p` potentials, currently only :math:`p=1`
+        is supported
+    :param neighbor_indices: torch.Tensor with the ``i,j`` indices of neighbors for
+        which the potential should be computed in real space.
+    :param neighbor_distances: torch.Tensor with the pair distances of the neighbors for
+        which the potential should be computed in real space.
+    :param ns_lo: Minimum number of spatial resolution along each axis
+    :param ns_hi: Maximum number of spatial resolution along each axis
+    :param accuracy: Recomended values for a balance between the accuracy and speed is
+        :math:`10^{-3}`. For more accurate results, use :math:`10^{-6}`.
+
+    :return: Tuple containing a float of the optimal smearing for the :class:
+        `CoulombPotential`, and a dictionary with the parameters for
+        :class:`EwaldCalculator`.
+
+    Example
+    -------
+    >>> import torch
+    >>> positions = torch.tensor(
+    ...     [[0.0, 0.0, 0.0], [0.4, 0.4, 0.4]], dtype=torch.float64
+    ... )
+    >>> charges = torch.tensor([[1.0], [-1.0]], dtype=torch.float64)
+    >>> cell = torch.eye(3, dtype=torch.float64)
+    >>> smearing, parameter = tune_ewald(
+    ...     charges, cell, positions, cutoff=4.4, accuracy=1e-1
+    ... )
+
+    """
+    _validate_parameters(charges, cell, positions, exponent)
+    min_dimension = float(torch.min(torch.linalg.norm(cell, dim=1)))
+    params = [{"lr_wavelength": min_dimension / ns} for ns in range(ns_lo, ns_hi + 1)]
+    tuner = GridSearchTuner(
+        charges,
+        cell,
+        positions,
+        cutoff,
```

Comment:
give maybe all as named paraeters?

```suggestion
        charges=charges,
        cell=cell,
        positions=positions,
        cutoff=cutoff,
```